// react-img-carousel/lib/controls/arrow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

/**
 * Renders an arrow component used to transition from right-to-left or left-to-right through the carousel slides.
 */
class Arrow extends _react.Component {
    static get propTypes() {
        return {
            numSlides: _propTypes.default.number.isRequired,
            selectedIndex: _propTypes.default.number.isRequired,
            infinite: _propTypes.default.bool.isRequired,
            prevSlide: _propTypes.default.func.isRequired,
            nextSlide: _propTypes.default.func.isRequired,
            direction: _propTypes.default.oneOf(['left', 'right']).isRequired,
            arrows: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.shape({
                left: _propTypes.default.node.isRequired,
                right: _propTypes.default.node.isRequired,
                className: _propTypes.default.string
            })])
        };
    }
    /**
     * @returns {Boolean} True if there is a next slide to transition to, else False.
     */


    hasNext() {
        const {
            direction,
            infinite,
            numSlides,
            selectedIndex
        } = this.props;
        return infinite || (direction === 'left' ? selectedIndex > 0 : selectedIndex < numSlides - 1);
    }

    render() {
        const {
            prevSlide,
            nextSlide,
            direction,
            arrows
        } = this.props;
        let arrowComponent = null;
        let buttonClass = 'carousel-arrow-default';

        if (arrows.left) {
            buttonClass = arrows.className ? arrows.className : '';
            arrowComponent = direction === 'left' ? arrows.left : arrows.right;
        }

        return /*#__PURE__*/ _react.default.createElement("button", {
            type: "button",
            disabled: !this.hasNext(),
            onClick: direction === 'left' ? prevSlide : nextSlide,
            className: `carousel-arrow carousel-${direction}-arrow ${buttonClass}`
        }, arrowComponent);
    }

}

exports.default = Arrow;
